pallets/automation-price/src/lib.rs

```diff
@@ -433,61 +467,54 @@ pub mod pallet {
-				let asset_sudo: AccountOf<T> = asset_metadatum.asset_sudo;
-				if asset_sudo != who {
-					Err(Error::<T>::InvalidAssetSudo)?
+			// TODO: ensure length are all same
+			for (index, price) in prices.clone().iter().enumerate() {
+				let index: usize = index.try_into().unwrap();
+
+				let chain = chains[index].clone();
+				let exchange = exchanges[index].clone();
+				let asset1 = assets1[index].clone();
+				let asset2 = assets2[index].clone();
+				let round = rounds[index].clone();
+
+				let key = (&chain, &exchange, (&asset1, &asset2));
+
+				if !AssetRegistry::<T>::contains_key(&key) {
+					// TODO: emit error and continue update the rest instead, temporary do this to
+					// keep going, will update later
+					Err(Error::<T>::AssetNotInitialized)?
 				}
-			}
-			let fee = <BalanceOf<T>>::saturated_from(1_000_000_000_000u64);
-			T::FeeHandler::can_pay_fee(&who, fee).map_err(|_| Error::<T>::InsufficientBalance)?;
-			if let Some(asset_target_price) = Self::get_asset_baseline_price(asset.clone()) {
-				let last_asset_price: AssetPrice = match Self::get_asset_price(asset.clone()) {
-					None => Err(Error::<T>::AssetNotSupported)?,
-					Some(asset_price) => asset_price,
-				};
-				let asset_update_percentage =
-					Self::calculate_asset_percentage(value, asset_target_price).saturating_add(1);
-				// NOTE: this is temporarily set to 0 for ease of calculation. Ideally, we can perform
-				// 			Self::calculate_asset_percentage(value, last_asset_price) and be able to compare the
-				// 			last percentage to the current one. However, calculate_asset_percentage does not return
-				// 			a direction. Therefore, let's say base price is 100. Last price is 95, current is 105.
-				// 			Since calculate_asset_percentage returns a u128, calculate_asset_percentage will return 5%
-				// 			for both, since there's no concept of positive/negative/direction (generic doesn't do just +/-).
-				// 			Therefore, this function will think 5% -> 5% means no change occurred, but instead we want to
-				// 			check 0% -> 5%. Therefore, we always check all the slots from 0% to x% where x is the updated
-				// 			percentage. This is less efficient, but more guaranteed. In the future, we will have to return
-				// 			direction for calculate_asset_percentage in the future.
-				let asset_last_percentage = 0;
-				if value > last_asset_price {
-					Self::move_scheduled_tasks(
-						asset.clone(),
-						asset_last_percentage,
-						asset_update_percentage,
-						Direction::Up,
-					)?;
-				} else {
-					Self::move_scheduled_tasks(
-						asset.clone(),
-						asset_last_percentage,
-						asset_update_percentage,
-						Direction::Down,
-					)?;
+
+				if let Some(asset_registry) = Self::get_asset_registry_info(key) {
+					let allow_wallets: Vec<AccountOf<T>> = asset_registry.oracle_providers;
+					if !allow_wallets.contains(&who) {
+						Err(Error::<T>::OracleNotAuthorized)?
+					}
+
+					// TODO: Add round and nonce check logic
```

Comment:
Is this to prevent re-entry attack? If yes, we should discuss our strategy here.

In regular transaction the nonce parameter is there to ensure same transaction is not submitted to chain twice. It’s basically a number based ID.